agent/consul/config_replication.go

```diff
@@ -0,0 +1,189 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+func cmpConfigLess(first structs.ConfigEntry, second structs.ConfigEntry) bool {
+	return first.GetKind() < second.GetKind() || (first.GetKind() == second.GetKind() && first.GetName() < second.GetName())
+}
+
+func configSort(configs []structs.ConfigEntry) {
+	sort.Slice(configs, func(i, j int) bool {
+		return cmpConfigLess(configs[i], configs[j])
+	})
+}
+
+func diffConfigEntries(local []structs.ConfigEntry, remote []structs.ConfigEntry, lastRemoteIndex uint64) ([]structs.ConfigEntry, []structs.ConfigEntry) {
+	configSort(local)
+	configSort(remote)
+
+	var deletions []structs.ConfigEntry
+	var updates []structs.ConfigEntry
+	var localIdx int
+	var remoteIdx int
+	for localIdx, remoteIdx = 0, 0; localIdx < len(local) && remoteIdx < len(remote); {
+		if local[localIdx].GetKind() == remote[remoteIdx].GetKind() && local[localIdx].GetName() == remote[remoteIdx].GetName() {
+			// config is in both the local and remote state - need to check raft indices and the Hash
+			if remote[remoteIdx].GetRaftIndex().ModifyIndex > lastRemoteIndex {
+				updates = append(updates, remote[remoteIdx])
+			}
+			// increment both indices when equal
+			localIdx += 1
+			remoteIdx += 1
+		} else if cmpConfigLess(local[localIdx], remote[remoteIdx]) {
+			// config no longer in remoted state - needs deleting
+			deletions = append(deletions, local[localIdx])
+
+			// increment just the local index
+			localIdx += 1
+		} else {
+			// local state doesn't have this config - needs updating
+			updates = append(updates, remote[remoteIdx])
+
+			// increment just the remote index
+			remoteIdx += 1
+		}
+	}
+
+	for ; localIdx < len(local); localIdx += 1 {
+		deletions = append(deletions, local[localIdx])
+	}
+
+	for ; remoteIdx < len(remote); remoteIdx += 1 {
+		updates = append(updates, remote[remoteIdx])
+	}
+
+	return deletions, updates
+}
+
+func (s *Server) reconcileLocalConfig(ctx context.Context, configs []structs.ConfigEntry, op structs.ConfigEntryOp) (bool, error) {
+	ticker := time.NewTicker(time.Second / time.Duration(s.config.ConfigReplicationApplyLimit))
+	defer ticker.Stop()
+
+	for i, entry := range configs {
+		req := structs.ConfigEntryRequest{
+			Op:         op,
+			Datacenter: s.config.Datacenter,
+			Entry:      entry,
+		}
+
+		resp, err := s.raftApply(structs.ConfigEntryRequestType, &req)
```

Comment:
Yeah, I was assuming a few things:

1. The rate of change to the centralized configs would be pretty low compared to ACLs.
2. The number of centralized config entries should also be much lower than the number of tokens + roles + policies.

Given those and the non-existence of batching capability currently I just did them singly.

It would be pretty easy to have the raft layer of the request only use batched operations. Thats is actually how ACLs work. Even when using the singular APIs the requests get translated to the final structure before being applied to raft. I quite like that but I was doing this task trying to work within the RPC/raft/state store infrastructure already in place. I could probably make this change pretty quickly and it would be better to do it now than have backwards compatibility issues in the future if it became a problem.

Reply:
I agree that we could do this later on if it became a problem.